src/functionalTest/java/uk/gov/hmcts/reform/sendletter/FunctionalTestSuite.java

```diff
@@ -0,0 +1,197 @@
+            .header(CONTENT_TYPE, APPLICATION_JSON_VALUE)
+            .baseUri(this.producerUrl)
+            .body(jsonBody)
+            .when()
+            .post("/letters")
+            .then()
+            .statusCode(200)
+            .extract()
+            .body()
+            .jsonPath()
+            .get("letter_id");
+    }
+
+    protected String sampleLetterRequestJson(int numberOfDocuments) throws JSONException {
+        return new JSONObject()
+            .put("type", SMOKE_TEST_LETTER_TYPE)
+            .put("documents", createDocumentJsonArray(numberOfDocuments))
+            .toString();
+    }
+
+    protected SFTPClient getSftpClient() throws IOException {
+        SSHClient ssh = new SSHClient();
+
+        ssh.addHostKeyVerifier(ftpFingerprint);
+        ssh.connect(ftpHostname, ftpPort);
+
+        ssh.authPublickey(
+            ftpUser,
+            ssh.loadKeys(ftpPrivateKey, ftpPublicKey, null)
+        );
+
+        return ssh.newSFTPClient();
+    }
+
+    protected ZipInputStream getZipInputStream(RemoteFile zipFile) throws IOException {
+        byte[] fileContent = new byte[(int) zipFile.length()];
+        zipFile.read(0, fileContent, 0, (int) zipFile.length());
+
+        ByteArrayInputStream inputStream =
+            new ByteArrayInputStream(fileContent, 0, fileContent.length);
+
+        return new ZipInputStream(inputStream);
+    }
+
+    protected byte[] readAllBytes(InputStream input) throws IOException {
+        ByteArrayOutputStream output = new ByteArrayOutputStream();
+        byte[] buffer = new byte[10000];
+        int len;
+
+        while ((len = input.read(buffer)) > 0) {
+            output.write(buffer, 0, len);
+        }
+
+        return output.toByteArray();
+    }
+
+    protected String getPdfFileNamePattern(String letterId) {
+        return String.format(
+            "%s_%s_%s.pdf",
+            Pattern.quote(SMOKE_TEST_LETTER_TYPE),
+            Pattern.quote(s2sName.replace("_", "")),
+            Pattern.quote(letterId)
+        );
+    }
+
+    protected String getZipFileNamePattern(String letterId) {
+        return String.format(
+            "%s_%s_\\d{14}_%s.zip",
+            Pattern.quote(SMOKE_TEST_LETTER_TYPE),
+            Pattern.quote(s2sName.replace("_", "")),
+            Pattern.quote(letterId)
+        );
+    }
+
+    @SuppressWarnings("PMD.AvoidInstantiatingObjectsInLoops")
+    private JSONArray createDocumentJsonArray(int numberOfDocuments) throws JSONException {
+        JSONArray jsonArray = new JSONArray();
+
+        for (int i = 0; i < numberOfDocuments; i++) {
+            jsonArray.put(new JSONObject()
+                .put("template", "<html>hello {{uuid}}</html>")
+                .put("values", new JSONObject()
+                    // so that we don't send the same letter twice
+                    .put("uuid", UUID.randomUUID().toString())
+                )
+            );
+        }
+
+        return jsonArray;
+    }
```

Comment:
we can avoid suppression by streaming up this process:

```java
Stream.generate(UUID::randomUUID)
    .map(uuid -> make that json object here)
    .limit(numberOfDocuments)
    .foreach(jsonArray::put)
```

Reply:
Not a bad idea, but the code that creates the JSON object throws a checked exception. Handling that exception would probably make the effort not worth it.

Reply:
ah bum. minor try/catch won't hurt

non-blocker really, i'll leave for you to decide:

| | pro? | con? |
|-|-|-|
|oldschool for loop| | |
|suppressed warning| | |
|stream usage| | |
|try/catch| | |

Reply:
Done